src/reducers/vis-state-merger.js

```diff
@@ -52,43 +51,49 @@ export function mergeFilters(state, filtersToMerge) {
 
   // merge filters
   filtersToMerge.forEach(filter => {
-    // match filter.dataId with current datesets id
-    // uploaded data need to have the same dataId with the filter
-    if (datasets[filter.dataId]) {
-      // datasets is already loaded
-      const validateFilter = validateFilterWithData(
-        datasets[filter.dataId],
-        filter
-      );
+    const datasetIds = Object.keys(datasets).reduce((acc, datasetId) => ([
+      ...acc,
+      ...(shouldApplyfilter(filter, datasetId) ? [datasetId] : [])
+    ]), []);
+
+    if (datasetIds.length) {
+
+      // for each dataset data mapped against the current filter are valid
+      const validatedFilter = datasetIds.reduce((acc, d) => {
```

Comment:
There is a scenario when 1 filter applies to multiple datasets ['a', 'b']. however, only 'a' is loaded at the moment. We should handle that scenario. in your current implementation, this filter will be merged with only 1 loaded dataset. I think we should send the filter to `unmerged` and wait until all datasets is loaded to merge it.



That's why I was asking what is the return value here

Reply:
We will apply a filter only if all datasets are present

Reply:
I updated the code a little bit and made sure we validate each filter against each dataset present in dataId